### DISCUSSION TITLE
Elsa V3: Facing issue registering custom persistent provider as Scoped

### DISCUSSION THREAD

**Discussion by viraj-ldce** [elsa-workflows/elsa-core#4107] [Q&A]
Hi,

I have integrated 3.0.0-rc1 packages for Elsa Server and I am trying to integrate my custom persistent provider. I am referring existing persistent providers to write my custom Stores. I used Elsa extension method **UseWorkflowManagement** to register my custom stores. For Workflow management, I have created Definition store and Instance Store and registering them with Elsa using below code.

```
public class MyWorkflowManagementPersistenceFeature : FeatureBase
{
	public MyWorkflowManagementPersistenceFeature(IModule module) : base(module)
	{
	}

	public override void Configure()
	{
		Module.Configure<WorkflowInstancesFeature>(feature => feature.WorkflowInstanceStore = sp => sp.GetRequiredService<MyInstanceStore>());
		Module.Configure<WorkflowDefinitionsFeature>(feature => feature.WorkflowDefinitionStore = sp => sp.GetRequiredService<MyDefinitionStore>());
	}

	public override void Apply()
	{
		base.Apply();

		Services.AddScoped<MyInstanceStore>();
		Services.AddScoped<MyDefinitionStore>();
	}
}
```

Now as I register them as ScopedService using .**AddScoped**, it's resulting into below error.
`System.InvalidOperationException: 'Cannot resolve scoped service 'POCElsaV3Server.MyStore.MyInstanceStore' from root provider.'`

I saw that built in Stores are registered as a Singleton. But I can not register my custom Stores as Singleton because these stores have dependency on other services which are scoped services.

How can I register custom persistence Stores as a Scoped and integrate them in Elsa ?
Why built in Stores are registered as a Singleton ? 

Regards,
Viraj

**Comment by sfmskywalker** [comment]
Hey Viraj,

Thanks for diving deep into Elsa Workflows and for your feedback.

**Why are built-in Stores registered as Singletons?**  
We went with singletons for our built-in stores mainly for efficiency. They're faster because we don't have to resolve these services all the time. Just to be clear, even though they're singletons, they're good citizens when it comes to database connections. They open them, do their thing, and close them quickly to avoid any connection issues.

**How to use Scoped persistence Stores in Elsa?**  
At the moment, Elsa isn't super friendly with registering the persistence stores as scoped services. That said, there's a workaround: inside your store, you can whip up a child scope using `IServiceScopeFactory` and get your persistence services from there. Here's a quick example:

```csharp
public class MyInstanceStore
{
    private readonly IServiceScopeFactory _scopeFactory;

    public MyInstanceStore(IServiceScopeFactory scopeFactory)
    {
        _scopeFactory = scopeFactory;
    }

    public Task SomeOperationAsync()
    {
        using var scope = _scopeFactory.CreateScope();
        
        var scopedService = scope.ServiceProvider.GetRequiredService<YourScopedService>();
        // do stuff with your scoped service
     }
}
```

Perhaps not the slickest solution, but it should do the trick for now. We'll definitely keep your feedback in mind for the next Elsa updates. Really appreciate you pointing it out!

Let me know if there's anything else or if you have more questions. Cheers!
